### PR TITLE
[Console] Add ReStructuredText descriptor

### LINKED ISSUE #45717: Add RST descriptor for console component.
### Description

The console component currently has descriptors for Markdown, Json, etc. This feature would add RST as an option.

### Example

    $helper = new DescriptorHelper();
    $helper->describe($output, $this->getApplication(), [
      'format' => 'rst',
    ]);

### PR DESCRIPTION
**Pull request by danepowell**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45717
| License       | MIT
| Doc PR        | n/a, descriptors are internal and not documented.

### Description

The console component currently has descriptors for Markdown, Json, etc. This feature would add Restructured Text descriptor (rst) as an option.

This replaces and addresses all feedback from #45718

### Example

```
$helper = new DescriptorHelper();
$helper->describe($output, $this->getApplication(), [
  'format' => 'rst',
]);
```

### CODE REVIEW

**Comment by danepowell** [comment]
I don't understand the Psalm failure. The error is `Uncaught Exception: InvalidArgumentException Could not get class storage for sessionhandlerinterface` which according to `--debug-by-line` comes from `/Users/dane.powell/src/symfony/src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MigratingSessionHandler.php`. I don't see how it could be related to this PR.

**Comment by danepowell** [comment]
@fabpot mind reviewing this since you approved the last one?

File: composer.json [hunk 1]
```diff
@@ -54,7 +54,8 @@
         "symfony/polyfill-intl-normalizer": "~1.0",
         "symfony/polyfill-mbstring": "~1.0",
         "symfony/polyfill-php83": "^1.27",
-        "symfony/polyfill-uuid": "^1.15"
+        "symfony/polyfill-uuid": "^1.15",
+        "ext-iconv": "*"
```
Comment (fabpot) [inline composer.json]: I would remove the dep here and use the String component instead when you've used `iconv`.

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+use Symfony\Component\Console\Helper\Helper;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputDefinition;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ReStructuredTextDescriptor extends Descriptor
+{
+    // <h1>
+    private string $partChar = '#';
```
Comment (fabpot) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: Let's be consistent with what we are using in the docs:

h1 =====
h2 -----
h3 ~~~~~
h4 .....

File: src/Symfony/Component/Console/Tests/Descriptor/AbstractDescriptorTest.php [hunk 3]
```diff
@@ -98,7 +98,9 @@ protected function assertDescription($expectedDescription, $describedObject, arr
     {
         $output = new BufferedOutput(BufferedOutput::VERBOSITY_NORMAL, true);
         $this->getDescriptor()->describe($output, $describedObject, $options + ['raw_output' => true]);
-        $this->assertEquals($this->normalizeOutput($expectedDescription), $this->normalizeOutput($output->fetch()));
+        $actual = $this->normalizeOutput($output->fetch());
+        $expected = $this->normalizeOutput($expectedDescription);
+        $this->assertEquals($expected, $actual);
```
Comment (fabpot) [inline src/Symfony/Component/Console/Tests/Descriptor/AbstractDescriptorTest.php]: Should be reverted as the code is semantically the same.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by danepowell** [comment]
Thanks for the review, all feedback addressed, this is good to go.

Tests are failing in HEAD, so probably not related to this PR. I think every PR I've opened has had failures of this nature.

Have you thought about using branch protections to prevent PRs / commits with failing tests from being merged?

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+            foreach ($commands as $command) {
+                $this->describeCommand($command, $options);
+                $this->write("\n\n");
+            }
+        }
+    }
+
+    /**
+     * @param \Symfony\Component\Console\Descriptor\ApplicationDescription $description
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: should be removed

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            if (!\in_array($option->getName(), $global_options)) {
+                $non_default_options[] = $option;
+            }
+        }
+
+        return $non_default_options;
+    }
+
+    /**
+     * @param \Symfony\Component\Console\Descriptor\ApplicationDescription $description
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: should be removed

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+     */
+    private function setVisibleNamespaces(ApplicationDescription $description): void
+    {
+        $commands = $description->getCommands();
+        foreach ($description->getNamespaces() as $namespace) {
+            try {
+                // Remove aliases.
+                $namespace_commands = $namespace['commands'];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: we use camel case: `$namespaceCommands`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                foreach ($namespace_commands as $key => $command_name) {
+                    if (!\array_key_exists($command_name, $commands)) {
+                        // If the array key does not exist, then this is an alias.
+                        unset($namespace_commands[$key]);
+                    } elseif ($commands[$command_name]->isHidden()) {
+                        unset($namespace_commands[$key]);
+                    }
+                }
+                if (!\count($namespace_commands)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `if (!$namespaceCommands) {`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Console\Descriptor;
+
+use Exception;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: should be replaced by inline `\Exception`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+                }
+            } catch (Exception) {
+            }
+            $this->visibleNamespaces[] = $namespace['id'];
+        }
+    }
+
+    private function removeAliasesAndHiddenCommands(array $commands): array
+    {
+        // Remove aliases.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: should be removed, the name of the method says it already

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+            'quiet',
+            'verbose',
+            'version',
+            'ansi',
+            'no-interaction',
+        ];
+        $non_default_options = [];
+        foreach ($definition->getOptions() as $option) {
+            // Skip global options.
+            // @todo Show these once at the beginning.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: let's do it?

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+            $this->write("\n\n");
+            $this->write(implode("\n", array_map(static function ($commandName) {
+                return sprintf('- `%s`_', $commandName);
+            }, array_keys($commands))));
+        }
+    }
+
+    private function getNonDefaultOptions(InputDefinition $definition): array
+    {
+        $global_options = [
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: camel case

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+            'version',
+            'ansi',
+            'no-interaction',
+        ];
+        $non_default_options = [];
+        foreach ($definition->getOptions() as $option) {
+            // Skip global options.
+            // @todo Show these once at the beginning.
+            if (!\in_array($option->getName(), $global_options)) {
+                $non_default_options[] = $option;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: camel case

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+                $commands = $application->all('');
+            } else {
+                $commands = $application->all($namespace);
+                $this->write("\n\n");
+                $this->write($namespace."\n".str_repeat($this->sectionChar, Helper::width($namespace))."\n\n");
+            }
+            $commands = $this->removeAliasesAndHiddenCommands($commands);
+
+            $this->write("\n\n");
+            $this->write(implode("\n", array_map(static function ($commandName) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `static fn ($commandName) => sprintf('- `%s`_', $commandName)`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+        $definition = $command->getDefinition();
+        if ($definition->getOptions() || $definition->getArguments()) {
+            $this->write("\n\n");
+            $this->describeInputDefinition($definition);
+        }
+    }
+
+    protected function describeApplication(Application $application, array $options = []): void
+    {
+        $describedNamespace = $options['namespace'] ?? null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: no need for the tmp variable

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+        $name = '\-\-'.$option->getName();
+        if ($option->isNegatable()) {
+            $name .= '|\-\-no-'.$option->getName();
+        }
+        if ($option->getShortcut()) {
+            $name .= '|-'.str_replace('|', '|-', $option->getShortcut());
+        }
+
+        $option_description = $option->getDescription() ? preg_replace('/\s*[\r\n]\s*/', "\n\n", $option->getDescription())."\n\n" : '';
+        $option_description = (new UnicodeString($option_description))->ascii();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: camel case

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+            .'- **Is value required**: '.($option->isValueRequired() ? 'yes' : 'no')."\n"
+            .'- **Is multiple**: '.($option->isArray() ? 'yes' : 'no')."\n"
+            .'- **Is negatable**: '.($option->isNegatable() ? 'yes' : 'no')."\n"
+            .'- **Default**: ``'.str_replace("\n", '', var_export($option->getDefault(), true)).'``'."\n"
+        );
+    }
+
+    protected function describeInputDefinition(InputDefinition $definition, array $options = []): void
+    {
+        if ($showArguments = (\count($definition->getArguments()) > 0)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `if ($showArguments = (bool) $definition->getArguments()) {`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+            $this->write("Arguments\n".str_repeat($this->subsubsectionChar, 9))."\n\n";
+            foreach ($definition->getArguments() as $argument) {
+                $this->write("\n\n");
+                $this->describeInputArgument($argument);
+            }
+        }
+
+        $non_default_options = $this->getNonDefaultOptions($definition);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: camel

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+        if (\count($non_default_options) > 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `if ($nonDefaultOptions = $this->getNonDefaultOptions($definition)) {`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+
+    protected function describeCommand(Command $command, array $options = []): void
+    {
+        if ($options['short'] ?? false) {
+            $this->write(
+                '``'.$command->getName()."``\n"
+                .str_repeat($this->subsectionChar, Helper::width($command->getName()))."\n\n"
+                .($command->getDescription() ? $command->getDescription()."\n\n" : '')
+                ."Usage\n".str_repeat($this->paragraphsChar, 5)."\n\n"
+                .array_reduce($command->getAliases(), static function ($carry, $usage) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `static fn ($carry, $usage) => ...`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+
+        foreach ($command->getAliases() as $alias) {
+            $this->write('.. _'.$alias.":\n\n");
+        }
+        $this->write(
+            $command->getName()."\n"
+            .str_repeat($this->subsectionChar, Helper::width($command->getName()))."\n\n"
+            .($command->getDescription() ? $command->getDescription()."\n\n" : '')
+            ."Usage\n".str_repeat($this->subsubsectionChar, 5)."\n\n"
+            .array_reduce(array_merge([$command->getSynopsis()], $command->getAliases(), $command->getUsages()), static function ($carry, $usage) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `fn (...) => ...`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+        $title = $this->getApplicationTitle($application);
+
+        $this->write($title."\n".str_repeat($this->partChar, Helper::width($title)));
+        $this->createTableOfContents($description, $application);
+        $this->describeCommands($application, $options);
+    }
+
+    private function getApplicationTitle(Application $application): string
+    {
+        if ('UNKNOWN' !== $application->getName()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]:
```php
if ('UNKNOWN' === $application->getName()) {
    return 'Console Tool';
}
```

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+            if ('_global' === $namespace) {
+                $commands = $application->all('');
+                $this->write('Global'."\n".str_repeat($this->sectionChar, Helper::width('Global'))."\n\n");
+            } else {
+                $commands = $application->all($namespace);
+                $this->write($namespace."\n".str_repeat($this->sectionChar, Helper::width($namespace))."\n\n");
+            }
+            $commands = $this->removeAliasesAndHiddenCommands($commands);
+
+            foreach ($commands as $command) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `foreach ($this->removeAliasesAndHiddenCommands($commands) as $command) {`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    /**
+     * @param \Symfony\Component\Console\Descriptor\ApplicationDescription $description
+     */
+    private function createTableOfContents(ApplicationDescription $description, Application $application): void
+    {
+        $this->setVisibleNamespaces($description);
+        $chapter_title = 'Table of Contents';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: camel

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+     * @param \Symfony\Component\Console\Descriptor\ApplicationDescription $description
+     */
+    private function setVisibleNamespaces(ApplicationDescription $description): void
+    {
+        $commands = $description->getCommands();
+        foreach ($description->getNamespaces() as $namespace) {
+            try {
+                // Remove aliases.
+                $namespace_commands = $namespace['commands'];
+                foreach ($namespace_commands as $key => $command_name) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `$commandName`

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+                        unset($namespace_commands[$key]);
+                    } elseif ($commands[$command_name]->isHidden()) {
+                        unset($namespace_commands[$key]);
+                    }
+                }
+                if (!\count($namespace_commands)) {
+                    // If the namespace contained only aliases or hidden commands, skip the namespace.
+                    continue;
+                }
+            } catch (Exception) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php]: `\Exception`

**Review by nicolas-grekas — COMMENTED** [review]
I only have CS fixes to suggest :)

**Comment by danepowell** [comment]
Thanks for the feedback, all comments addressed.

**Comment by anavarre** [comment]
Is there anything else missing here?

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @danepowell.

**Comment by fabpot** [comment]
@danepowell Would you be interested in supporting restructuredtext in `src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/` as well?

**Comment by danepowell** [comment]
I only build console apps and actively avoid the Framework bundle because of all the web-specific bulk. So I don't have an interest in adding to it.

**Pull request closed without merging**